Add mod_infinite_loop option

alexhenrie opened the pull request:
This is definitely the largest new feature that I have tried to add to Uncrustify. To keep it as simple as possible, `mod_infinite_loop` is pretty strict about what input it accepts and what output it creates. I welcome any feedback you may have!

micheleCTDE commented:
Hi @alexhenrie,
I will have a look at this during the weekend and feedback. Sorry for the wait.

micheleCTDE commented on src/log_levels.h:
```diff
@@ -132,6 +132,7 @@ enum log_sev_t
    LOUTPUT   = 101, //! output
    LUNC      = 102, //! rules used in uncrustify.cpp
    LQT       = 103, //! load/save options for Qt
+   LMINFLOOP = 104, //! mod_infinite_loop
```
I think we don't need a specific log option for this, unless you can explain a good use case for it. Using 'log_rule_B' should suffice, but happy to hear @guy-maurel 's opinion as well.

alexhenrie replied:
I have removed the LOG_FMT calls as requested.

micheleCTDE commented on src/rewrite_infinite_loops.cpp:
```diff
@@ -0,0 +1,328 @@
+
+static bool for_needs_rewrite(Chunk *pc, E_Token desired_type)
+{
+   assert(chunk_is_token(pc, CT_FOR));
```
better to use 'if' rather than assert and return if the condition is not met, IMO. Likewise in the rest of the file.

micheleCTDE commented on src/rewrite_infinite_loops.cpp:
```diff
@@ -0,0 +1,328 @@
+      break;
+
+   default:
+      assert(false);
```
Don't 'assert', simply print a log message about the wrong value and return without doing any changes.

micheleCTDE commented on src/rewrite_infinite_loops.cpp:
```diff
@@ -0,0 +1,328 @@
+} // while_needs_rewrite
+
+
+static void move_one_token(Chunk **source, Chunk **destination, E_Token parent_type)
```
There is a Swap method in Chunk. Could we use that to swap to chunk perhaps? Just suggesting to have a look, maybe what you need here is more that what the function does.

micheleCTDE replied:
ok, makes sense. It was just a thought I wanted to share.

micheleCTDE reviewed (changes requested):
Some things to clean up, but it seems promising.
Would be better if it was a bit less stringent, allowing separators to be accepted too, like for example
```for ( ; ; )``` or ```while ( true )```

alexhenrie commented on tests/expected/c/10081-infinite-loops.c:
```diff
@@ -0,0 +1,39 @@
+{
+	int x;
+
+	for (;;)
```
I'm expecting the user to set `sp_before_sparen`, `sp_inside_sparen`, and `sp_inside_for` in addition to `mod_infinite_loop`. Otherwise, there is no guarantee that loops will have similar spacing. Do you want me to set `sp_before_sparen` in the tests so that all the loops in the expected output look the same?

micheleCTDE replied:
Ah, I see now. The ones with the additional space are those cases which were not replaced and so are the same as the original code. In such case I think it should be fine, because other options should take care of normalizing spacing later on. I have not tested, but if you have verified that doing so is enough, then we can just ignore my original comment.

alexhenrie commented on src/chunk.cpp:
```diff
@@ -471,6 +471,7 @@ void Chunk::Delete(Chunk * &pc)
 void Chunk::MoveAfter(Chunk *ref)
 {
    LOG_FUNC_ENTRY();
+   assert(ref != this);
```
Done.

alexhenrie commented:
@micheleCTDE Thanks for the feedback. I have tried to address all of your comments, but I may need a little more direction to understand what you'd like to see.

micheleCTDE commented:
> @micheleCTDE Thanks for the feedback. I have tried to address all of your comments, but I may need a little more direction to understand what you'd like to see.

I will take a look this weekend and comment back. Apologies for the wait, it's another busy week...

micheleCTDE commented on tests/expected/c/10082-infinite-loops.c:
```diff
@@ -0,0 +1,39 @@
+	{
+	}
+
+	while (true)
```
as per comment above

micheleCTDE commented on tests/expected/c/10083-infinite-loops.c:
```diff
@@ -0,0 +1,43 @@
+	do
+		x++; while(true);
+
+	do
+		x++; while(true);
+
+	do
+		x++;
+	while (true);
```
If settings other options results in a consistent indentation/spacing, that would be fine. But if that results in do-while loops with different formatting in the end, then we should consider doing some extra work here.

micheleCTDE replied:
ok, seems reasonable

micheleCTDE commented on src/rewrite_infinite_loops.cpp:
```diff
@@ -0,0 +1,310 @@
+      }
+   }
+
+   if (!chunk_is_token(keyword, desired_type))
+   {
+      return(true);
+   }
```
This seems wrong. If the result is true, it says it is ok to rewrite, without checking that the condition is "true" or "1".

alexhenrie replied:
You're right; when changing from `while` to `for`, the function should still make sure that the original condition was `true` or `1`. I have now fixed that and added a test for it.

micheleCTDE replied:
great

micheleCTDE commented on src/rewrite_infinite_loops.cpp:
```diff
@@ -0,0 +1,310 @@
+{
+   Chunk *next_source = (*source)->GetNext();
+
+   // Place the token immediately after the previous token, without any
+   // whitespace.
```
Comment should refer to source and destination, which not necessarily are adjacent.

alexhenrie replied:
I have rephrased the comment to use the words "source" and "destination".

micheleCTDE commented on src/rewrite_infinite_loops.cpp:
```diff
@@ -0,0 +1,310 @@
+   // whitespace.
+
+   (*source)->MoveAfter(*destination);
+   (*source)->column       = (*destination)->column + (*destination)->Len();
+   (*source)->orig_col     = (*destination)->orig_col + (*destination)->Len();
+   (*source)->orig_col_end = (*source)->orig_col_end + (*source)->Len();
+   (*source)->orig_prev_sp = 0;
```
I have not done a real test, but once 'source' is moved after 'destination', do the destination's column/orig_col/orig_col_end/orig_prev_sp also need some updating? why only source's values are modified? if I am missing something obvious, please let me know.

micheleCTDE replied:
Still worth to check and make sure the values in the dumps are correct, even if uncrustify sorts things out fine in the end.
You can use the -ds option to see intermiediate steps and see if there is any inconsistent column values.

alexhenrie replied:
I tried the `-ds` option as you suggested and it created six parse dumps, numbered `-c_002.log` through `-c_007.log`. The tokens are shown as overlapping in `-c_002.log` through `-c_004.log`. However, in `-c_005.log` through `-c_007.log`, the tokens are moved so that they do not overlap. I am not sure how to solve this, and I'm still not sure that it's a real problem. Perhaps we can leave it as it is for now, and revisit it later if it turns out to be important?

`infinite-loops.c`:

```
	do // comment
		x++;
	while (1); // comment
```

`-c_002.log`:

```
# 113>                FOR|               NONE|     PARENT_NOT_SET[  9/  9/ 12/  0][1/1/0][        c 0000][0-1]         for
# 115>        SPAREN_OPEN|                FOR|     PARENT_NOT_SET[ 12/ 12/ 14/  0][1/1/0][   2 0000 0000][0-0]            (
# 115>          SEMICOLON|                FOR|     PARENT_NOT_SET[ 13/ 13/ 16/  0][1/2/0][        8 0020][0-0]             ;
# 115>          SEMICOLON|                FOR|     PARENT_NOT_SET[ 13/ 13/ 16/  0][1/2/0][        8 0020][0-0]             ;
# 115>       SPAREN_CLOSE|                FOR|     PARENT_NOT_SET[ 14/ 14/ 17/  0][1/1/0][   2 0000 0000][0-0]              )
# 113>        VBRACE_OPEN|                 DO|     PARENT_NOT_SET[ 12/  9/  0/  0][1/1/0][             0][0-0]
# 113>        COMMENT_CPP|        COMMENT_END|     PARENT_NOT_SET[ 12/ 12/ 22/  1][2/2/0][             0][0-0]            // comment
# 113>            NEWLINE|               NONE|     PARENT_NOT_SET[ 22/ 22/ 17/  0][2/2/0][             0][1-0]
# 114>               WORD|               NONE|     PARENT_NOT_SET[ 17/ 17/ 18/  0][2/2/0][        c 0000][0-1]                 x
# 114>       INCDEC_AFTER|               NONE|     PARENT_NOT_SET[ 18/ 18/ 20/  0][2/2/0][   2 0000 0000][0-0]                  ++
# 114>          SEMICOLON|               NONE|     PARENT_NOT_SET[ 20/ 20/ 21/  0][2/2/0][   2 0000 0000][0-0]                    ;
# 114>       VBRACE_CLOSE|                 DO|     PARENT_NOT_SET[  0/ 20/  0/  0][1/1/0][             0][0-0]
# 114>            NEWLINE|               NONE|     PARENT_NOT_SET[ 21/ 21/  9/  0][1/1/0][             0][1-0]
# 115>        COMMENT_CPP|        COMMENT_END|     PARENT_NOT_SET[ 20/ 20/ 30/  1][1/1/0][             0][0-0]                    // comment
```

`-c_005.log`:

```
# 113>                FOR|               NONE|     PARENT_NOT_SET[  9/  9/ 12/  0][1/1/0][        c 0000][0-1]         for
# 115>        SPAREN_OPEN|                FOR|     PARENT_NOT_SET[ 12/ 12/ 14/  0][1/1/0][   2 0000 0000][0-0]            (
# 115>          SEMICOLON|                FOR|     PARENT_NOT_SET[ 13/ 13/ 16/  0][1/2/0][        8 0020][0-0]             ;
# 115>          SEMICOLON|                FOR|     PARENT_NOT_SET[ 14/ 13/ 16/  0][1/2/0][        8 0020][0-0]              ;
# 115>       SPAREN_CLOSE|                FOR|     PARENT_NOT_SET[ 15/ 14/ 17/  0][1/1/0][   2 0000 0000][0-0]               )
# 113>        VBRACE_OPEN|                 DO|     PARENT_NOT_SET[ 16/  9/  0/  0][1/1/0][             0][0-0]
# 113>        COMMENT_CPP|        COMMENT_END|     PARENT_NOT_SET[ 17/ 12/ 22/  1][2/2/0][             0][0-0]                 // comment
# 113>            NEWLINE|               NONE|     PARENT_NOT_SET[ 27/ 22/ 17/  0][2/2/0][             0][1-0]
# 114>               WORD|               NONE|     PARENT_NOT_SET[ 17/ 17/ 18/  0][2/2/0][        c 0000][0-1]                 x
# 114>       INCDEC_AFTER|               NONE|     PARENT_NOT_SET[ 18/ 18/ 20/  0][2/2/0][   2 0000 0000][0-0]                  ++
# 114>          SEMICOLON|               NONE|     PARENT_NOT_SET[ 20/ 20/ 21/  0][2/2/0][   2 0000 0000][0-0]                    ;
# 114>       VBRACE_CLOSE|                 DO|     PARENT_NOT_SET[ 22/ 20/  0/  0][1/1/0][             0][0-0]
# 114>            NEWLINE|               NONE|     PARENT_NOT_SET[ 22/ 21/  9/  0][1/1/0][             0][1-0]
# 115>        COMMENT_CPP|      COMMENT_WHOLE|     PARENT_NOT_SET[  9/ 20/ 30/  1][1/1/0][             0][0-0]         // comment
```

micheleCTDE commented:
@alexhenrie 
Sorry for the long delay. Please see my comments above, some changes are likely to be required.
As we progress with refining this PR, I will do some actual testing (not done yet), so I may have more comments in future.

micheleCTDE commented:
Forgot to mention that it would be good to add some negative tests, that is tests where loops don't get replaced because they don't meet the required/allowed syntax. This will help to catch eventual mistakes.

alexhenrie commented:
> Forgot to mention that it would be good to add some negative tests, that is tests where loops don't get replaced because they don't meet the required/allowed syntax. This will help to catch eventual mistakes.

Thanks for the feedback. I have added several more test cases, including some related to the problem you found where the loop condition was not being checked. Do you see anything else amiss?

micheleCTDE commented on tests/expected/c/10085-infinite-loops.c:
```diff
@@ -0,0 +1,83 @@
+	while (1);
+
+	do
+		x++; while(1);
```
I think we could add a new like before ```while(1)``` when replacing into do-while loop. That should be an easy change and probably what everyone would expect, regardless of what settings is used for other options.

micheleCTDE commented on tests/expected/c/10083-infinite-loops.c:
```diff
@@ -0,0 +1,83 @@
+	while(true);
+
+	do
+		x++; while(true);
```
I think we could add a new like before while(true) when replacing into do-while loop. That should be an easy change and probably what everyone would expect, regardless of what settings is used for other options.

micheleCTDE commented on src/rewrite_infinite_loops.cpp:
```diff
@@ -0,0 +1,316 @@
+   }
+   pc = pc->GetNext();
+
+   if (!chunk_is_token(pc, CT_SPAREN_OPEN))
```
you can replace ```chunk_is_token()``` with ```->Is()``` since we have started to do this in the code.

alexhenrie replied:
Done

micheleCTDE reviewed (changes requested):
Seems ready in terms of functionality. Few minor code improvements (see comments).
After that I will do a test of a ~3200 file repo and if no regression are highlighted, we can pass this to @guy-maurel for his comments and final approval.

alexhenrie commented on src/rewrite_infinite_loops.cpp:
```diff
@@ -0,0 +1,316 @@
+}
+
+
+static void rewrite_loop_condition(Chunk **source, Chunk **destination,
```
Done

micheleCTDE replied:
Great!

alexhenrie commented:
Hi Michele, I have addressed your comments above and added a bunch more test cases just to be sure. The output does look ugly in a few places, but as we discussed, the user can easily get prettier output by setting more options. I hope that the code is good enough to move forward now.

micheleCTDE commented:
> Hi Michele, I have addressed your comments above and added a bunch more test cases just to be sure. The output does look ugly in a few places, but as we discussed, the user can easily get prettier output by setting more options. I hope that the code is good enough to move forward now.

Hi @alexhenrie,
thanks for the patience and the adjustments. Seems good to go.
The output does look a bit ugly at times, I agree. But setting other options should take care of it as discussed.

I could be more picky here and say for example that a
```
for (;;) {
...
}
```
should look like 
```
do {
...
}
while (true);
```
for example. This will result in less ugly replacements.

@guy-maurel as maintainer, what is your opinion? Any other comment about the PR overall?

alexhenrie commented:
> I could be more picky here and say for example that a
> 
> ```
> for (;;) {
> ...
> }
> ```
> 
> should look like
> 
> ```
> do {
> ...
> }
> while (true);
> ```
> 
> for example. This will result in less ugly replacements. Similar observations for other replacements, the test files have some good example of things that could be better formatted.

That would be a good future improvement. I considered trying to preserve the original formatting as much as possible (like you are suggesting), but I decided that it was too complicated to do as part of the initial pull request. What we have now is a good start and although it's not absolutely perfect, it works well and would be a very helpful feature for me.

micheleCTDEAdmin commented:
> @guy-maurel as maintainer, what is your opinion? Any other comment about the PR overall?

Guy has confirmed to me that he is happy with this PR, so I will merge it.

micheleCTDEAdmin commented:
@alexhenrie thanks for the good work!

alexhenrie commented:
Thank you!
